File: py/bank.py
from typing import Tuple
from money import Money


class Bank:
    def __init__(self) -> None:
        self._exchange_rates = {}

    def add_exchange_rate(self, currency_from: str, currency_to: str, rate: float) -> None:
        exchange = self._build_exchange_name(currency_from, currency_to)
        self._exchange_rates[exchange] = rate

    def convert(self, money: Money, currency: str) -> Tuple:
        if money.currency == currency:
            return Money(money.amount, currency), None

        exchange = self._build_exchange_name(money.currency, currency)
        if exchange in self._exchange_rates:
            return (Money(money.amount * self._exchange_rates[exchange], currency), None)

        return (None, exchange)

    def _build_exchange_name(self, currency_from: str, currency_to: str) -> str:
        return f'{currency_from}->{currency_to}'

File: js/bank.js
import Money from './money.js';

class Bank {
  constructor() {
    this.exchangeRates = new Map();
  }

  addExchangeRate(currencyFrom, currencyTo, rate) {
    const exchange = this._buildExchangeName(currencyFrom, currencyTo);
    this.exchangeRates.set(exchange, rate);
  }

  convert(money, currency) {
    if (money.currency === currency) {
      return new Money(money.amount, money.currency);
    }

    const exchange = this._buildExchangeName(money.currency, currency);
    const rate = this.exchangeRates.get(exchange);

    if (!rate) {
      throw new Error(exchange);
    }
    
    return new Money(money.amount * rate, currency);
  }

  _buildExchangeName(currencyFrom, currencyTo) {
    return `${currencyFrom}->${currencyTo}`;
  }
}

export default Bank;
